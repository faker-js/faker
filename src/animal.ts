import type { Faker } from '.';

export class Animal {
  constructor(private readonly faker: Faker) {
    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Animal.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }
  }

  /**
   * Returns a random dog breed.
   *
   */
  dog(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.dog);
  }

  /**
   * Returns a random cat breed.
   *
   */
  cat(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cat);
  }

  /**
   * Returns a random snake species.
   *
   */
  snake(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.snake);
  }

  /**
   * Returns a random bear species.
   *
   */
  bear(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bear);
  }

  /**
   * Returns a random lion species.
   *
   */
  lion(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.lion);
  }

  /**
   * Returns a random cetacean species.
   *
   * @method faker.animal.cetacean
   */
  cetacean(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.cetacean
    );
  }

  /**
   * Returns a random horse breed.
   *
   */
  horse(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.horse);
  }

  /**
   * Returns a random bird species.
   *
   */
  bird(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bird);
  }

  /**
   * Returns a random cow species.
   *
   */
  cow(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cow);
  }

  /**
   * Returns a random fish species.
   *
   */
  fish(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.fish);
  }

  /**
   * Returns a random crocodilian species.
   *
   */
  crocodilia(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.crocodilia
    );
  }

  /**
   * Returns a random insect species.
   *
   */
  insect(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.insect);
  }

  /**
   * Returns a random rabbit species
   *
   */
  rabbit(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.rabbit);
  }

  /**
   * Returns a random animal type.
   *
   */
  type(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.type);
  }
}
