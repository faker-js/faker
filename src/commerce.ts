import type { Faker } from '.';
import { deprecated } from './internal/deprecated';

/**
 * Module to generate commerce and product related entries.
 */
export class Commerce {
  constructor(private readonly faker: Faker) {
    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Commerce.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }
  }

  /**
   * Returns a human readable color name.
   *
   * @example
   * faker.commerce.color() // 'red'
   *
   * @deprecated
   */
  color(): string {
    deprecated({
      deprecated: 'faker.commerce.color()',
      proposed: 'faker.color.human()',
      since: 'v7.0.0',
      until: 'v8.0.0',
    });
    return this.faker.color.human();
  }

  /**
   * Returns a department inside a shop.
   *
   * @example
   * faker.commerce.department() // 'Garden'
   */
  department(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.commerce.department
    );
  }

  /**
   * Return a percentage value in decimal format between 0 and 1
   * with percision of two decimal place.
   *
   * @param precision Precision of the generated number. Defaults to `0.01`.
   *
   * @example
   * this.getPercentage(0.1) // 0.4
   * this.getPercentage(0.01) // 0.36
   */
  percentage(precision: number): number {
    return this.faker.datatype.float({
      min: 0,
      max: 1,
      precision: precision || 0.01,
    });
  }

  /**
   * Generates a random descriptive product name.
   *
   * @example
   * faker.commerce.productName() // 'Incredible Soft Gloves'
   */
  productName(): string {
    return (
      this.faker.commerce.productAdjective() +
      ' ' +
      this.faker.commerce.productMaterial() +
      ' ' +
      this.faker.commerce.product()
    );
  }

  /**
   * Generates a price between min and max (inclusive).
   *
   * @param min The minimum price. Defaults to `1`.
   * @param max The maximum price. Defaults to `1000`.
   * @param dec The number of decimal places. Defaults to `2`.
   * @param symbol The currency value to use. Defaults to `''`.
   *
   * @example
   * faker.commerce.price() // 828.00
   * faker.commerce.price(100) // 904.00
   * faker.commerce.price(100, 200) // 154.00
   * faker.commerce.price(100, 200, 0) // 133
   * faker.commerce.price(100, 200, 0, '$') // $114
   */
  price(
    min: number = 1,
    max: number = 1000,
    dec: number = 2,
    symbol: string = ''
  ): string {
    if (min < 0 || max < 0) {
      return `${symbol}${0.0}`;
    }

    const randValue = this.faker.datatype.number({ max: max, min: min });

    return (
      symbol +
      (Math.round(randValue * Math.pow(10, dec)) / Math.pow(10, dec)).toFixed(
        dec
      )
    );
  }

  /**
   * Returns an adjective describing a product.
   *
   * @example
   * faker.commerce.productAdjective() // 'Handcrafted'
   */
  productAdjective(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.commerce.product_name.adjective
    );
  }

  /**
   * Returns a material of a product.
   *
   * @example
   * faker.commerce.productMaterial() // 'Rubber'
   */
  productMaterial(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.commerce.product_name.material
    );
  }

  /**
   * Returns a short product name.
   *
   * @example
   * faker.commerce.product() // 'Computer'
   */
  product(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.commerce.product_name.product
    );
  }

  /**
   * Returns a product description.
   *
   * @example
   * faker.commerce.productDescription() // 'Andy shoes are designed to keeping...'
   */
  productDescription(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.commerce.product_description
    );
  }
}
