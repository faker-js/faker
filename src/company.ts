import type { Faker } from '.';
import type { Fake } from './fake';

let f: Fake['fake'];

export class Company {
  constructor(private readonly faker: Faker) {
    f = this.faker.fake;

    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Company.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }
  }

  /**
   * suffixes
   *
   * @method faker.company.suffixes
   */
  suffixes(): string[] {
    // Don't want the source array exposed to modification, so return a copy
    return this.faker.definitions.company.suffix.slice(0);
  }

  /**
   * companyName
   *
   * @method faker.company.companyName
   * @param format
   */
  companyName(format?: number): string {
    const formats = [
      '{{name.lastName}} {{company.companySuffix}}',
      '{{name.lastName}} - {{name.lastName}}',
      '{{name.lastName}}, {{name.lastName}} and {{name.lastName}}',
    ];

    if (typeof format !== 'number') {
      format = this.faker.datatype.number(formats.length - 1);
    }

    return f(formats[format]);
  }

  /**
   * companySuffix
   *
   * @method faker.company.companySuffix
   */
  companySuffix(): string {
    return this.faker.random.arrayElement(this.faker.company.suffixes());
  }

  /**
   * catchPhrase
   *
   * @method faker.company.catchPhrase
   */
  catchPhrase(): string {
    return f(
      '{{company.catchPhraseAdjective}} {{company.catchPhraseDescriptor}} {{company.catchPhraseNoun}}'
    );
  }

  /**
   * bs
   *
   * @method faker.company.bs
   */
  bs(): string {
    return f('{{company.bsBuzz}} {{company.bsAdjective}} {{company.bsNoun}}');
  }

  /**
   * catchPhraseAdjective
   *
   * @method faker.company.catchPhraseAdjective
   */
  catchPhraseAdjective(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.company.adjective
    );
  }

  /**
   * catchPhraseDescriptor
   *
   * @method faker.company.catchPhraseDescriptor
   */
  catchPhraseDescriptor(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.company.descriptor
    );
  }

  /**
   * catchPhraseNoun
   *
   * @method faker.company.catchPhraseNoun
   */
  catchPhraseNoun(): string {
    return this.faker.random.arrayElement(this.faker.definitions.company.noun);
  }

  /**
   * bsAdjective
   *
   * @method faker.company.bsAdjective
   */
  bsAdjective(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.company.bs_adjective
    );
  }

  /**
   * bsBuzz
   *
   * @method faker.company.bsBuzz
   */
  bsBuzz(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.company.bs_verb
    );
  }

  /**
   * bsNoun
   *
   * @method faker.company.bsNoun
   */
  bsNoun(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.company.bs_noun
    );
  }
}
