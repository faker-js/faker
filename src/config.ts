/**
 * The possible configuration options, that can be set.
 * This type exists to be extended for plugins via type augmentation.
 *
 * The `@default` tag is used to indicate the default value, that should be used if, the config is absent.
 */
export interface FakerConfig {
  /**
   * The function used to generate the `refDate` date instance, if not provided as method param.
   * The function must return a new valid `Date` instance for every call.
   *
   * @see [Reproducible Results](https://fakerjs.dev/guide/usage.html#reproducible-results)
   * @see faker.seed(): For generating reproducible values.
   *
   * @since 9.0.0
   *
   * @default () => new Date()
   */
  defaultRefDate?: () => Date;
}
