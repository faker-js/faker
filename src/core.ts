import type { FakerConfig } from './config';
import type { LocaleDefinition } from './definitions';
import type { Randomizer } from './randomizer';

/**
 * The core grants access to the locale data, the randomizer and config settings.
 */
export interface FakerCore {
  /**
   * The locale data associated with this instance.
   *
   * Always present, but it might be empty if the locale data is not available.
   */
  readonly locale: LocaleDefinition;

  /**
   * The randomizer used to generate random values.
   */
  readonly randomizer: Randomizer;

  /**
   * The configuration settings used by this instance.
   */
  readonly config: FakerConfig;
}
