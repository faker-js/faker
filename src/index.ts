import { Faker } from './faker';
import allLocales from './locales';

export type {
  AnimalDefinitions,
  ColorDefinitions,
  CommerceDefinitions,
  CommerceProductNameDefinitions,
  CompanyDefinitions,
  DatabaseDefinitions,
  DateDefinitions,
  DateEntryDefinition,
  FinanceCurrencyEntryDefinitions,
  FinanceDefinitions,
  HackerDefinitions,
  InternetDefinitions,
  LocaleDefinition,
  /** @deprecated Use LocationDefinitions instead */
  LocationDefinitions as AddressDefinitions,
  LocationDefinitions,
  LoremDefinitions,
  MusicDefinitions,
  /** @deprecated Use PersonDefinitions instead */
  PersonDefinitions as NameDefinitions,
  PersonDefinitions,
  /** @deprecated Use PersonTitleDefinitions instead */
  PersonTitleDefinitions as NameTitleDefinitions,
  PersonTitleDefinitions,
  PhoneNumberDefinitions,
  ScienceDefinitions,
  SystemDefinitions,
  SystemMimeTypeEntryDefinitions,
  VehicleDefinitions,
  WordDefinitions,
} from './definitions';
export { FakerError } from './errors/faker-error';
export type { FakerOptions, UsableLocale, UsedLocales } from './faker';
export type { AnimalModule } from './modules/animal';
export type {
  Casing,
  ColorFormat,
  ColorModule,
  CSSFunction,
  CSSSpace,
  NumberColorFormat,
  StringColorFormat,
} from './modules/color';
export type { CommerceModule } from './modules/commerce';
export type { CompanyModule } from './modules/company';
export type { DatabaseModule } from './modules/database';
export type { DatatypeModule } from './modules/datatype';
export type { DateModule } from './modules/date';
export type { FinanceModule } from './modules/finance';
export type { GitModule } from './modules/git';
export type { HackerModule } from './modules/hacker';
export type { HelpersModule } from './modules/helpers';
export type { ImageModule } from './modules/image';
export type { InternetModule } from './modules/internet';
export type {
  /** @deprecated Use LocationModule instead */
  LocationModule as AddressModule,
  LocationModule,
} from './modules/location';
export type { LoremModule } from './modules/lorem';
export type { MusicModule } from './modules/music';
export { Sex } from './modules/person';
export type {
  /** @deprecated Use PersonModule instead */
  PersonModule as NameModule,
  PersonModule,
  SexType,
} from './modules/person';
export type { PhoneModule } from './modules/phone';
export type { RandomModule } from './modules/random';
export type { ChemicalElement, ScienceModule, Unit } from './modules/science';
export type { StringModule } from './modules/string';
export type { SystemModule } from './modules/system';
export type { VehicleModule } from './modules/vehicle';
export type { WordModule } from './modules/word';
export { Faker };

// since we are requiring the top level of faker, load all locales by default
export const faker: Faker = new Faker({
  locales: allLocales,
});
