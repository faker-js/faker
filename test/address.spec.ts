import { afterEach, describe, expect, it } from 'vitest';
import { faker } from '../src';
import { seededTests } from './support/seededRuns';
import { times } from './support/times';

function degreesToRadians(degrees: number) {
  return degrees * (Math.PI / 180.0);
}

function kilometersToMiles(miles: number) {
  return miles * 0.621371;
}

// http://nssdc.gsfc.nasa.gov/planetary/factsheet/earthfact.html
const EQUATORIAL_EARTH_RADIUS = 6378.137;

function haversine(
  latitude1: number,
  longitude1: number,
  latitude2: number,
  longitude2: number,
  isMetric: boolean
) {
  const distanceLatitude = degreesToRadians(latitude2 - latitude1);
  const distanceLongitude = degreesToRadians(longitude2 - longitude1);
  const a =
    Math.sin(distanceLatitude / 2) * Math.sin(distanceLatitude / 2) +
    Math.cos(degreesToRadians(latitude1)) *
      Math.cos(degreesToRadians(latitude2)) *
      Math.sin(distanceLongitude / 2) *
      Math.sin(distanceLongitude / 2);
  const distance =
    EQUATORIAL_EARTH_RADIUS * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

  return isMetric ? distance : kilometersToMiles(distance);
}

const NON_SEEDED_BASED_RUN = 5;

describe('address', () => {
  afterEach(() => {
    faker.locale = 'en';
  });

  seededTests(faker, 'address', (t) => {
    t.itEach('street', 'streetName', 'streetPrefix', 'streetSuffix');

    t.it('buildingNumber');

    t.it('secondaryAddress');
    t.describe('streetAddress', (t) => {
      t.it('noArgs')
        .it('with useFullAddress = true', true)
        .it('with useFullAddress = false', false);
    });

    t.it('cityName')
      .it('cityPrefix')
      .it('citySuffix')
      .describe('city', (t) => {
        t.it('noArgs').it('with given index', 1);
      });

    t.it('county');

    t.it('country').describe('countryCode', (t) => {
      t.it('noArgs')
        .it('with code = alpha-2', 'alpha-2')
        .it('with code = alpha-3', 'alpha-3');
    });

    t.describe('latitude', (t) => {
      t.it('noArgs');
    });
    t.describe('longitude', (t) => {
      t.it('noArgs');
    });

    t.describe('nearbyGPSCoordinate', (t) => {
      t.it('noArgs').it('near origin', [0, 0]);
    });
    t.it('state').it('stateAbbr');

    t.it('timeZone');

    t.describeEach(
      'direction',
      'cardinalDirection',
      'ordinalDirection'
    )((t) => {
      t.it('noArgs')
        .it('with abbr = true', true)
        .it('with abbr = false', false);
    });

    t.describe('zipCode', (t) => {
      t.it('noArgs').it('with format', '###-###');
    });

    t.describe('zipCodeByState', (t) => {
      t.it('state', 'CA');
      t.it('state2', 'WA');
    });
  });

  describe(`random seeded tests for seed ${faker.seed()}`, () => {
    for (let i = 1; i <= NON_SEEDED_BASED_RUN; i++) {
      describe('countryCode()', () => {
        it('returns random alpha-3 countryCode', () => {
          const countryCode = faker.address.countryCode('alpha-3');

          expect(countryCode).toBeTruthy();
          expect(
            countryCode.length,
            'The countryCode should be 3 characters long'
          ).toBe(3);
        });
      });

      describe('zipCode()', () => {
        it('returns random zipCode - user specified format', () => {
          let zipCode = faker.address.zipCode('?#? #?#');

          expect(zipCode).toMatch(/^[A-Za-z]\d[A-Za-z]\s\d[A-Za-z]\d$/);

          // try another format
          zipCode = faker.address.zipCode('###-###');

          expect(zipCode).toMatch(/^\d{3}-\d{3}$/);
        });

        it('returns zipCode with proper locale format', () => {
          // we'll use the en_CA locale..
          faker.locale = 'en_CA';
          const zipCode = faker.address.zipCode();

          expect(zipCode).toMatch(/^[A-Za-z]\d[A-Za-z]\s?\d[A-Za-z]\d$/);
        });
      });

      describe('zipCodeByState()', () => {
        it('returns zipCode valid for specified State', () => {
          faker.locale = 'en_US';
          const states = ['IL', 'GA', 'WA'];

          const zipCode1 = +faker.address.zipCodeByState(states[0]);
          expect(zipCode1).toBeGreaterThanOrEqual(60001);
          expect(zipCode1).toBeLessThanOrEqual(62999);

          const zipCode2 = +faker.address.zipCodeByState(states[1]);
          expect(zipCode2).toBeGreaterThanOrEqual(30001);
          expect(zipCode2).toBeLessThanOrEqual(31999);

          const zipCode3 = +faker.address.zipCodeByState(states[2]);
          expect(zipCode3).toBeGreaterThanOrEqual(98001);
          expect(zipCode3).toBeLessThanOrEqual(99403);
        });
      });

      describe('latitude()', () => {
        it('returns random latitude', () => {
          for (let i = 0; i < 100; i++) {
            const latitude = faker.address.latitude();

            expect(latitude).toBeTypeOf('string');

            const latitude_float = parseFloat(latitude);

            expect(latitude_float).toBeGreaterThanOrEqual(-90.0);
            expect(latitude_float).toBeLessThanOrEqual(90.0);
          }
        });

        it('returns latitude with min and max and default precision', () => {
          for (let i = 0; i < 100; i++) {
            const latitude = faker.address.latitude(5, -5);

            expect(latitude).toBeTypeOf('string');
            expect(
              latitude.split('.')[1].length,
              'The precision of latitude should be 4 digits'
            ).toBe(4);

            const latitude_float = parseFloat(latitude);

            expect(latitude_float).toBeGreaterThanOrEqual(-5);
            expect(latitude_float).toBeLessThanOrEqual(5);
          }
        });

        it('returns random latitude with custom precision', () => {
          for (let i = 0; i < 100; i++) {
            const latitude = faker.address.latitude(undefined, undefined, 7);

            expect(latitude).toBeTypeOf('string');
            expect(
              latitude.split('.')[1].length,
              'The precision of latitude should be 7 digits'
            ).toBe(7);

            const latitude_float = parseFloat(latitude);

            expect(latitude_float).toBeGreaterThanOrEqual(-180);
            expect(latitude_float).toBeLessThanOrEqual(180);
          }
        });
      });

      describe('longitude()', () => {
        it('returns random longitude', () => {
          for (let i = 0; i < 100; i++) {
            const longitude = faker.address.longitude();

            expect(longitude).toBeTypeOf('string');

            const longitude_float = parseFloat(longitude);

            expect(longitude_float).toBeGreaterThanOrEqual(-180);
            expect(longitude_float).toBeLessThanOrEqual(180);
          }
        });

        it('returns random longitude with min and max and default precision', () => {
          for (let i = 0; i < 100; i++) {
            const longitude = faker.address.longitude(100, -30);

            expect(longitude).toBeTypeOf('string');
            expect(
              longitude.split('.')[1].length,
              'The precision of longitude should be 4 digits'
            ).toBe(4);

            const longitude_float = parseFloat(longitude);

            expect(longitude_float).toBeGreaterThanOrEqual(-30);
            expect(longitude_float).toBeLessThanOrEqual(100);
          }
        });

        it('returns random longitude with custom precision', () => {
          for (let i = 0; i < 100; i++) {
            const longitude = faker.address.longitude(undefined, undefined, 7);

            expect(longitude).toBeTypeOf('string');
            expect(
              longitude.split('.')[1].length,
              'The precision of longitude should be 7 digits'
            ).toBe(7);

            const longitude_float = parseFloat(longitude);

            expect(longitude_float).toBeGreaterThanOrEqual(-180);
            expect(longitude_float).toBeLessThanOrEqual(180);
          }
        });
      });

      describe('direction()', () => {
        it('returns abbreviation when useAbbr is true', () => {
          const direction = faker.address.direction(true);
          const lengthDirection = direction.length;
          const prefixErrorMessage =
            'The abbreviation of direction when useAbbr is true should';

          expect(
            direction,
            `${prefixErrorMessage} be of type string. Current is ${typeof direction}`
          ).toBeTypeOf('string');
          expect(lengthDirection).toBeLessThanOrEqual(2);
        });
      });

      describe('ordinalDirection()', () => {
        it('returns abbreviation when useAbbr is true', () => {
          const ordinalDirection = faker.address.ordinalDirection(true);
          const expectedType = 'string';
          const ordinalDirectionLength = ordinalDirection.length;
          const prefixErrorMessage =
            'The ordinal direction when useAbbr is true should';

          expect(
            ordinalDirection,
            `${prefixErrorMessage} be equal ${expectedType}. Current is ${typeof ordinalDirection}`
          ).toBeTypeOf(expectedType);
          expect(ordinalDirectionLength).toBeLessThanOrEqual(2);
        });
      });

      describe('cardinalDirection()', () => {
        it('returns abbreviation when useAbbr is true', () => {
          const cardinalDirection = faker.address.cardinalDirection(true);
          const expectedType = 'string';
          const cardinalDirectionLength = cardinalDirection.length;
          const prefixErrorMessage =
            'The cardinal direction when useAbbr is true should';

          expect(
            cardinalDirection,
            `${prefixErrorMessage} be of type ${expectedType}. Current is ${typeof cardinalDirection}`
          ).toBeTypeOf(expectedType);
          expect(cardinalDirectionLength).toBeLessThanOrEqual(2);
        });
      });

      describe('nearbyGPSCoordinate()', () => {
        for (const isMetric of [true, false]) {
          for (const radius of times(100)) {
            it.each(times(5))(
              `should return random gps coordinate within a distance of another one (${JSON.stringify(
                { isMetric, radius }
              )}) (iter: %s)`,
              () => {
                const latitude1 = +faker.address.latitude();
                const longitude1 = +faker.address.longitude();

                const coordinate = faker.address.nearbyGPSCoordinate(
                  [latitude1, longitude1],
                  radius,
                  isMetric
                );

                expect(coordinate.length).toBe(2);
                expect(coordinate[0]).toBeTypeOf('string');
                expect(coordinate[1]).toBeTypeOf('string');

                const latitude2 = +coordinate[0];
                expect(latitude2).toBeGreaterThanOrEqual(-90.0);
                expect(latitude2).toBeLessThanOrEqual(90.0);

                const longitude2 = +coordinate[1];
                expect(longitude2).toBeGreaterThanOrEqual(-180.0);
                expect(longitude2).toBeLessThanOrEqual(180.0);

                const actualDistance = haversine(
                  latitude1,
                  longitude1,
                  latitude2,
                  longitude2,
                  isMetric
                );
                expect(actualDistance).toBeLessThanOrEqual(radius);
              }
            );
          }
        }
      });
    }
  });
});
