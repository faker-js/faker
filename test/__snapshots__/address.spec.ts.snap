// Vitest Snapshot v1

exports[`address > 42 > buildingNumber 1`] = `"7917"`;

exports[`address > 42 > cardinalDirection > noArgs 1`] = `"East"`;

exports[`address > 42 > cardinalDirection > with abbr = false 1`] = `"East"`;

exports[`address > 42 > cardinalDirection > with abbr = true 1`] = `"E"`;

exports[`address > 42 > city > noArgs 1`] = `"Port Valentine"`;

exports[`address > 42 > city > with given index 1`] = `"West Peyton"`;

exports[`address > 42 > cityName 1`] = `"Gulfport"`;

exports[`address > 42 > cityPrefix 1`] = `"West"`;

exports[`address > 42 > citySuffix 1`] = `"bury"`;

exports[`address > 42 > country 1`] = `"Haiti"`;

exports[`address > 42 > countryCode > noArgs 1`] = `"GY"`;

exports[`address > 42 > countryCode > with code = alpha-2 1`] = `"GY"`;

exports[`address > 42 > countryCode > with code = alpha-3 1`] = `"SAU"`;

exports[`address > 42 > county 1`] = `"Berkshire"`;

exports[`address > 42 > direction > noArgs 1`] = `"South"`;

exports[`address > 42 > direction > with abbr = false 1`] = `"South"`;

exports[`address > 42 > direction > with abbr = true 1`] = `"S"`;

exports[`address > 42 > latitude > noArgs 1`] = `"-22.5828"`;

exports[`address > 42 > longitude > noArgs 1`] = `"-45.1656"`;

exports[`address > 42 > nearbyGPSCoordinate > near origin 1`] = `
[
  "0.0814",
  "-0.0809",
]
`;

exports[`address > 42 > nearbyGPSCoordinate > noArgs 1`] = `
[
  "-22.5828",
  "106.7555",
]
`;

exports[`address > 42 > ordinalDirection > noArgs 1`] = `"Northwest"`;

exports[`address > 42 > ordinalDirection > with abbr = false 1`] = `"Northwest"`;

exports[`address > 42 > ordinalDirection > with abbr = true 1`] = `"NW"`;

exports[`address > 42 > secondaryAddress 1`] = `"Apt. 791"`;

exports[`address > 42 > state 1`] = `"Maine"`;

exports[`address > 42 > stateAbbr 1`] = `"ME"`;

exports[`address > 42 > street 1`] = `"Peyton Village"`;

exports[`address > 42 > streetAddress > noArgs 1`] = `"7917 Miller Pine"`;

exports[`address > 42 > streetAddress > with useFullAddress = false 1`] = `"7917 Miller Pine"`;

exports[`address > 42 > streetAddress > with useFullAddress = true 1`] = `"7917 Miller Pine Apt. 410"`;

exports[`address > 42 > streetName 1`] = `"Peyton Village"`;

exports[`address > 42 > streetPrefix 1`] = `"b"`;

exports[`address > 42 > streetSuffix 1`] = `"Isle"`;

exports[`address > 42 > timeZone 1`] = `"Europe/Amsterdam"`;

exports[`address > 42 > zipCode > noArgs 1`] = `"79177"`;

exports[`address > 42 > zipCode > with format 1`] = `"379-177"`;

exports[`address > 42 > zipCodeByState > state 1`] = `"79177"`;

exports[`address > 42 > zipCodeByState > state2 1`] = `"79177"`;

exports[`address > 1211 > buildingNumber 1`] = `"487"`;

exports[`address > 1211 > cardinalDirection > noArgs 1`] = `"West"`;

exports[`address > 1211 > cardinalDirection > with abbr = false 1`] = `"West"`;

exports[`address > 1211 > cardinalDirection > with abbr = true 1`] = `"W"`;

exports[`address > 1211 > city > noArgs 1`] = `"La Crosse"`;

exports[`address > 1211 > city > with given index 1`] = `"Fort Jadyn"`;

exports[`address > 1211 > cityName 1`] = `"Urbana"`;

exports[`address > 1211 > cityPrefix 1`] = `"Fort"`;

exports[`address > 1211 > citySuffix 1`] = `"shire"`;

exports[`address > 1211 > country 1`] = `"Uganda"`;

exports[`address > 1211 > countryCode > noArgs 1`] = `"UM"`;

exports[`address > 1211 > countryCode > with code = alpha-2 1`] = `"UM"`;

exports[`address > 1211 > countryCode > with code = alpha-3 1`] = `"IRN"`;

exports[`address > 1211 > county 1`] = `"Cambridgeshire"`;

exports[`address > 1211 > direction > noArgs 1`] = `"Southwest"`;

exports[`address > 1211 > direction > with abbr = false 1`] = `"Southwest"`;

exports[`address > 1211 > direction > with abbr = true 1`] = `"SW"`;

exports[`address > 1211 > latitude > noArgs 1`] = `"77.1337"`;

exports[`address > 1211 > longitude > noArgs 1`] = `"154.2673"`;

exports[`address > 1211 > nearbyGPSCoordinate > near origin 1`] = `
[
  "-0.0287",
  "0.0596",
]
`;

exports[`address > 1211 > nearbyGPSCoordinate > noArgs 1`] = `
[
  "77.1337",
  "-14.7545",
]
`;

exports[`address > 1211 > ordinalDirection > noArgs 1`] = `"Southwest"`;

exports[`address > 1211 > ordinalDirection > with abbr = false 1`] = `"Southwest"`;

exports[`address > 1211 > ordinalDirection > with abbr = true 1`] = `"SW"`;

exports[`address > 1211 > secondaryAddress 1`] = `"Suite 487"`;

exports[`address > 1211 > state 1`] = `"Washington"`;

exports[`address > 1211 > stateAbbr 1`] = `"WA"`;

exports[`address > 1211 > street 1`] = `"Koelpin Turnpike"`;

exports[`address > 1211 > streetAddress > noArgs 1`] = `"487 Breana Wells"`;

exports[`address > 1211 > streetAddress > with useFullAddress = false 1`] = `"487 Breana Wells"`;

exports[`address > 1211 > streetAddress > with useFullAddress = true 1`] = `"487 Breana Wells Apt. 616"`;

exports[`address > 1211 > streetName 1`] = `"Koelpin Turnpike"`;

exports[`address > 1211 > streetPrefix 1`] = `"c"`;

exports[`address > 1211 > streetSuffix 1`] = `"Via"`;

exports[`address > 1211 > timeZone 1`] = `"Asia/Magadan"`;

exports[`address > 1211 > zipCode > noArgs 1`] = `"48721-9061"`;

exports[`address > 1211 > zipCode > with format 1`] = `"948-721"`;

exports[`address > 1211 > zipCodeByState > state 1`] = `"48721-9061"`;

exports[`address > 1211 > zipCodeByState > state2 1`] = `"48721-9061"`;

exports[`address > 1337 > buildingNumber 1`] = `"51225"`;

exports[`address > 1337 > cardinalDirection > noArgs 1`] = `"East"`;

exports[`address > 1337 > cardinalDirection > with abbr = false 1`] = `"East"`;

exports[`address > 1337 > cardinalDirection > with abbr = true 1`] = `"E"`;

exports[`address > 1337 > city > noArgs 1`] = `"New Carmella"`;

exports[`address > 1337 > city > with given index 1`] = `"West Kellen"`;

exports[`address > 1337 > cityName 1`] = `"Dubuque"`;

exports[`address > 1337 > cityPrefix 1`] = `"West"`;

exports[`address > 1337 > citySuffix 1`] = `"boro"`;

exports[`address > 1337 > country 1`] = `"Equatorial Guinea"`;

exports[`address > 1337 > countryCode > noArgs 1`] = `"EH"`;

exports[`address > 1337 > countryCode > with code = alpha-2 1`] = `"EH"`;

exports[`address > 1337 > countryCode > with code = alpha-3 1`] = `"PRI"`;

exports[`address > 1337 > county 1`] = `"Bedfordshire"`;

exports[`address > 1337 > direction > noArgs 1`] = `"South"`;

exports[`address > 1337 > direction > with abbr = false 1`] = `"South"`;

exports[`address > 1337 > direction > with abbr = true 1`] = `"S"`;

exports[`address > 1337 > latitude > noArgs 1`] = `"-42.8356"`;

exports[`address > 1337 > longitude > noArgs 1`] = `"-85.6711"`;

exports[`address > 1337 > nearbyGPSCoordinate > near origin 1`] = `
[
  "0.0806",
  "-0.0061",
]
`;

exports[`address > 1337 > nearbyGPSCoordinate > noArgs 1`] = `
[
  "-42.8356",
  "21.7907",
]
`;

exports[`address > 1337 > ordinalDirection > noArgs 1`] = `"Northwest"`;

exports[`address > 1337 > ordinalDirection > with abbr = false 1`] = `"Northwest"`;

exports[`address > 1337 > ordinalDirection > with abbr = true 1`] = `"NW"`;

exports[`address > 1337 > secondaryAddress 1`] = `"Apt. 512"`;

exports[`address > 1337 > state 1`] = `"Indiana"`;

exports[`address > 1337 > stateAbbr 1`] = `"IN"`;

exports[`address > 1337 > street 1`] = `"Kellen Dam"`;

exports[`address > 1337 > streetAddress > noArgs 1`] = `"51225 Alexys Haven"`;

exports[`address > 1337 > streetAddress > with useFullAddress = false 1`] = `"51225 Alexys Haven"`;

exports[`address > 1337 > streetAddress > with useFullAddress = true 1`] = `"51225 Alexys Haven Apt. 552"`;

exports[`address > 1337 > streetName 1`] = `"Kellen Dam"`;

exports[`address > 1337 > streetPrefix 1`] = `"a"`;

exports[`address > 1337 > streetSuffix 1`] = `"Forks"`;

exports[`address > 1337 > timeZone 1`] = `"Africa/Casablanca"`;

exports[`address > 1337 > zipCode > noArgs 1`] = `"51225"`;

exports[`address > 1337 > zipCode > with format 1`] = `"251-225"`;

exports[`address > 1337 > zipCodeByState > state 1`] = `"51225"`;

exports[`address > 1337 > zipCodeByState > state2 1`] = `"51225"`;
